Timeout in streamRecv of client

ashwin95r opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/gru/19)
<!-- Reviewable:end -->

pawanrawal commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[client/main.go, line 31 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLodSDE_TmK3rMZjFTI:-KLodSDF_7VNQ-6wPFYY:-1856495250) ([raw file](https://github.com/dgraph-io/gru/blob/ce96fd849e095fb7d5f5cefe9c9f17753b3e0427/client/main.go#L31)):*
> ```Go
> var address = flag.String("address", "54.183.229.38:50001", "Address of the server")
> var endTT chan *interact.ServerStatus
> var timeoutChan chan int
> ```

Don't have it as a global variable, pass it to your go routine. Also is it possible to encapsulate endTT within any other struct that is there?

---

*[client/main.go, line 120 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLodDNaEVVT2eHYIXCz:-KLodDNbYJxS122Tpige:770331045) ([raw file](https://github.com/dgraph-io/gru/blob/ce96fd849e095fb7d5f5cefe9c9f17753b3e0427/client/main.go#L120)):*
> ```Go
> 	case <-timeoutChan:
> 		return
> 	case <-time.After(11 * time.Second):
> ```

How did you arrive at this value? Does this trigger in case the net connection is not available at the client?

---

*[client/main.go, line 128 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLodGBL7Oy9dg94Fm9d:-KLodGBL7Oy9dg94Fm9e:-1859226153) ([raw file](https://github.com/dgraph-io/gru/blob/ce96fd849e095fb7d5f5cefe9c9f17753b3e0427/client/main.go#L128)):*
> ```Go
> 
> func streamRecv(stream interact.GruQuiz_StreamChanClient) {
> 	timeoutChan = make(chan int, 1)
> ```

This should be named better. Probably it should be msgChan as it indicates message was received. We don't pass value to it when timeout occurs. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[client/main.go, line 131 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLoeWHu7FZ_e3nwkr2X:-KLoeWHu7FZ_e3nwkr2Y:-1123426863) ([raw file](https://github.com/dgraph-io/gru/blob/ce96fd849e095fb7d5f5cefe9c9f17753b3e0427/client/main.go#L131)):*
> ```Go
> 	for {
> 		go checkTimeout()
> 		msg, err := stream.Recv()
> ```

Have a goroutine which receives from stream, does the error checking and passes it onto a channel. Then, do a select over two channels, one from this channel, and another using time.After(2 * duration of pings) in an infinite for loop.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[client/main.go, line 122 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLogBbBKGPZVos9R8bk:-KLogBbCSSe9AMrlCjNG:-198854482) ([raw file](https://github.com/dgraph-io/gru/blob/ce96fd849e095fb7d5f5cefe9c9f17753b3e0427/client/main.go#L122)):*
> ```Go
> 	case <-time.After(11 * time.Second):
> 		log.Println("No pings received from server")
> 		showErrorPage()
> ```

Don't show a complete page for this issue. Take one last line in the client and always show the status of connection. If it's healthy show the last ping received from server at X time, status: OK. And if you don't receive pings anymore, then say connection issue with the server, using red fonts or something.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 1 of 2 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[client/main.go, line 31 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLodSDE_TmK3rMZjFTI:-KLp0qL49OCFn0LzKfU6:-145748624) ([raw file](https://github.com/dgraph-io/gru/blob/ce96fd849e095fb7d5f5cefe9c9f17753b3e0427/client/main.go#L31)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Don't have it as a global variable, pass it to your go routine. Also is it possible to encapsulate endTT within any other struct that is there?
</details>
Moved it into functions.

---

*[client/main.go, line 120 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLodDNaEVVT2eHYIXCz:-KLoe6Qk-cRiE9igJwC9:-973246116) ([raw file](https://github.com/dgraph-io/gru/blob/ce96fd849e095fb7d5f5cefe9c9f17753b3e0427/client/main.go#L120)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> How did you arrive at this value? Does this trigger in case the net connection is not available at the client?
</details>
Yeah, it triggers 

---

*[client/main.go, line 122 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLogBbBKGPZVos9R8bk:-KLou33RER735Yy4KtJ1:-458014693) ([raw file](https://github.com/dgraph-io/gru/blob/ce96fd849e095fb7d5f5cefe9c9f17753b3e0427/client/main.go#L122)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Don't show a complete page for this issue. Take one last line in the client and always show the status of connection. If it's healthy show the last ping received from server at X time, status: OK. And if you don't receive pings anymore, then say connection issue with the server, using red fonts or something.
</details>
Yeah, We need to add a box for this purpose​, would make for a separate feature/PR.

---

*[client/main.go, line 128 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLodGBL7Oy9dg94Fm9d:-KLozESQlJlXie8okoAx:-609445857) ([raw file](https://github.com/dgraph-io/gru/blob/ce96fd849e095fb7d5f5cefe9c9f17753b3e0427/client/main.go#L128)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> This should be named better. Probably it should be msgChan as it indicates message was received. We don't pass value to it when timeout occurs. 
</details>
Changed.

---

*[client/main.go, line 131 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLoeWHu7FZ_e3nwkr2X:-KLozakFHO9xgPTMlJvm:1715478026) ([raw file](https://github.com/dgraph-io/gru/blob/ce96fd849e095fb7d5f5cefe9c9f17753b3e0427/client/main.go#L131)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Have a goroutine which receives from stream, does the error checking and passes it onto a channel. Then, do a select over two channels, one from this channel, and another using time.After(2 * duration of pings) in an infinite for loop.
</details>
Okay, made some restructuring. Does it look alright now?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Review status: 1 of 2 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[client/main.go, line 120 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp5YurLp3Gz_DeVpGG:-KLp5YurLp3Gz_DeVpGH:1702124650) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L120)):*
> ```Go
> 		gotPing <- 1
> 		if err != nil {
> 			if err != io.EOF {
> ```

Is this check even relevant? We are doing the same thing 
```
endTT <- msg
return
```
 
irrespective of this check. Also end of test is denoted by msg.Status= "END" or an error while receiving ping?

---

*[client/main.go, line 121 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp5ALfCQHmv05Hu4fL:-KLp5ALfCQHmv05Hu4fM:1667426820) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L121)):*
> ```Go
> 		if err != nil {
> 			if err != io.EOF {
> 				//log.Printf("Error while receiving stream, %v", err)
> ```

Maybe remove this if not used.

---

*[client/main.go, line 154 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp45fQvvNlYIqq9_iv:-KLp45fR60M498WYJw57:1996137117) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L154)):*
> ```Go
> 		case <-gotPing:
> 			continue
> 		case <-time.After(10 * time.Second):
> ```

Have a variable for ping duration. And do 2*pingDuration here.

---

*[client/main.go, line 169 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp6FZTDGNh6qJhc4In:-KLp6FZTDGNh6qJhc4Io:-1433888755) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L169)):*
> ```Go
> 		select {
> 		case _ = <-endTT:
> 			return
> ```

Assuming that we get a value on endTT when the server is down, shouldn't we show an error to the user here?

---

*[client/main.go, line 199 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp2sRMuPnkHKU37ZcZ:-KLp2sRNk04iKBCvxDQ-:-771728193) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L199)):*
> ```Go
> 	}
> 
> 	cliStat := &interact.ClientStatus{
> ```

Going through the whole code for streaming.Do we need this and the `stream.Send` here? One thing is that `"First"` isn't an actual qid.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 1 of 2 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[client/main.go, line 120 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp5YurLp3Gz_DeVpGG:-KLpCBipz6kV1N_mPkVF:1558349486) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L120)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Is this check even relevant? We are doing the same thing 
> ```
> endTT <- msg
> return
> ```
>  
> irrespective of this check. Also end of test is denoted by msg.Status= "END" or an error while receiving ping?
</details>
True. In this case we do the same on both. But in general, if err is io.EOF, it means the stream has been closed.

---

*[client/main.go, line 121 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp5ALfCQHmv05Hu4fL:-KLpCWy27yXhRW-j1zmV:1974651005) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L121)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Maybe remove this if not used.
</details>
Now we'll have a common log for all errors.

---

*[client/main.go, line 154 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp45fQvvNlYIqq9_iv:-KLpDzVl2ybea9Lx18La:-499146009) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L154)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Have a variable for ping duration. And do 2*pingDuration here.
</details>
Done.

---

*[client/main.go, line 169 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp6FZTDGNh6qJhc4In:-KLpD9C1XlF8hXdX3Ixg:-58200380) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L169)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Assuming that we get a value on endTT when the server is down, shouldn't we show an error to the user here?
</details>
It will be logged when we insert an element to the channel.

---

*[client/main.go, line 199 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp2sRMuPnkHKU37ZcZ:-KLpEqaWTN4A24-6bpwX:1868156936) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L199)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Going through the whole code for streaming.Do we need this and the `stream.Send` here? One thing is that `"First"` isn't an actual qid.
</details>
Ah yeah, this was needed before we had the authenticate RPC when the token had to be exchanged. It's not needed now. Removing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[client/main.go, line 169 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp6FZTDGNh6qJhc4In:-KLpsaTk6X-Quyn8Lapp:-1275628856) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L169)):*
<details><summary><i>Previously, ashwin95r (Ashwin Ramesh) wrote…</i></summary>
> It will be logged when we insert an element to the channel.
</details>
Ok, when you get an error you aren't taking any action. I have a feeling that we can make do with one out of endTT and gotPing. See if that's possible.

---

*[client/main.go, line 120 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLprSMKZ1381nRUBU-q:-KLprSMKZ1381nRUBU-r:-742978331) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L120)):*
> ```Go
> 		gotPing <- 1
> 		if err != nil {
> 			endTT <- msg
> ```

If there is an error then won't msg be nil? 

---

*[client/main.go, line 187 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpr7N9VBusWPT73DdN:-KLpr7NA57w0exXJTIMT:-1015141782) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L187)):*
> ```Go
> 
> func initializeStream() {
> 	endTT := make(chan *interact.ServerStatus, 1)
> ```

Maybe name this endTest or better errorChan because you use this to send error values, its more understandable that way.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[client/main.go, line 117 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpvWKvNww-f58jqK80:-KLpvWKvNww-f58jqK81:746521873) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L117)):*
> ```Go
> 	endTT chan *interact.ServerStatus) {
> 	for {
> 		msg, err := stream.Recv()
> ```

It's not clear what you're doing with msg.

---

*[client/main.go, line 118 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpubFBGP3QSOpjmfAb:-KLpubFBGP3QSOpjmfAc:-1326065390) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L118)):*
> ```Go
> 	for {
> 		msg, err := stream.Recv()
> 		gotPing <- 1
> ```

Why do you need gotPing?

---

*[client/main.go, line 143 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpwQQPG2hImqQv1_fN:-KLpwQQPG2hImqQv1_fO:-317654678) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L143)):*
> ```Go
> func streamRecv(stream interact.GruQuiz_StreamChanClient,
> 	endTT chan *interact.ServerStatus) {
> 	gotPing := make(chan int, 1)
> ```

Why have such a small buffer? It can lead to blocking.

---

*[client/main.go, line 152 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpurQXdz-gdl56efUu:-KLpurQXdz-gdl56efUv:-631124412) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L152)):*
> ```Go
> 			log.Println("No pings received from server")
> 			showErrorPage()
> 			return
> ```

return? Isn't that going to stop the test? That's not the right behavior. The point is to keep on retrying.

---

*[client/main.go, line 158 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpwF2h6diOO1mju8we:-KLpwF2h6diOO1mju8wf:-183723440) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L158)):*
> ```Go
> 
> func streamSend(stream interact.GruQuiz_StreamChanClient,
> 	endTT chan *interact.ServerStatus) {
> ```

if endTT is only for tracking error, why not endTT chan error?

---

*[client/main.go, line 166 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpvAh7y36ViLskCA0l:-KLpvAh7y36ViLskCA0m:1430786075) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L166)):*
> ```Go
> 			return
> 		case <-tickChan:
> 			{
> ```

Do you need this block?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 1 of 2 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[client/main.go, line 169 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLp6FZTDGNh6qJhc4In:-KLqUzULSbAInKFqR-mC:2144686422) ([raw file](https://github.com/dgraph-io/gru/blob/038bdea17ff922388c229bcd5974061a1b5f0933/client/main.go#L169)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Ok, when you get an error you aren't taking any action. I have a feeling that we can make do with one out of endTT and gotPing. See if that's possible.
</details>
errChan and gotPing are to communicate with two different go-rountines. One is to signal the streamSend() and other for streamRecv()

---

*[client/main.go, line 117 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpvWKvNww-f58jqK80:-KLqTXXI1Dx6fOm2lqrQ:-1260448330) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L117)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> It's not clear what you're doing with msg.
</details>
msg has the information about how much time is left in the test and if the test is over, which is denoted by an "END" token.

---

*[client/main.go, line 118 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpubFBGP3QSOpjmfAb:-KLqTgu5s6wH7jN-GKIB:1767200298) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L118)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Why do you need gotPing?
</details>
This is to signal the parent go -routine that we got a ping

---

*[client/main.go, line 143 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpwQQPG2hImqQv1_fN:-KLqUBOFF4JsIo0KGjvp:-1935635380) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L143)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Why have such a small buffer? It can lead to blocking.
</details>
gotPing should ideally have only one element in it and it should immediately be consumed by the select statement that is blocking on gotPing and the time.After() call.

---

*[client/main.go, line 152 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpurQXdz-gdl56efUu:-KLqSB0ySMF4HypIjPUR:1700266662) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L152)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> return? Isn't that going to stop the test? That's not the right behavior. The point is to keep on retrying.
</details>
It won't stop the test,  just returns from this go-routine/stream. Currently, if this happens once, we show an error box and ask them to mail the logfile. Once we have the box to show the connection status, we would not return and just display a message saying temporary connection problem.

---

*[client/main.go, line 158 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpwF2h6diOO1mju8we:-KLqUxJSqZJ4rY1cZCZ6:-499146009) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L158)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> if endTT is only for tracking error, why not endTT chan error?
</details>
Done.

---

*[client/main.go, line 166 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpvAh7y36ViLskCA0l:-KLqT2XTUwL-P5C2lhUd:-431016890) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L166)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> Do you need this block?
</details>
This is the block that sends the current question being viewed by the client every 5 seconds.

---

*[client/main.go, line 187 \[r3\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLpr7N9VBusWPT73DdN:-KLqUS_BL-OeEA9m3XGg:-1158000865) ([raw file](https://github.com/dgraph-io/gru/blob/90c185b2ab80c5ce7b84d217c6e1743349e494bc/client/main.go#L187)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Maybe name this endTest or better errorChan because you use this to send error values, its more understandable that way.
</details>
Sure, if that removes some confusion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[client/main.go, line 129 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLtlt4j25H7Wkfp7LRi:-KLtlt4j25H7Wkfp7LRj:1039407749) ([raw file](https://github.com/dgraph-io/gru/blob/82a2575cbbe34ea89251bd2bebfc096eca06f9bb/client/main.go#L129)):*
> ```Go
> 
> func streamRecv(stream interact.GruQuiz_StreamChanClient) {
> 	gotPing := make(chan *interact.ServerStatus)
> ```

It's an unbuffered channel. This would definitely lead to blocking. If you're not sure why is it blocking, read the channels chapter in the book. You'll gain a good understanding of channels now that you're playing with them.

---

*[client/main.go, line 176 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLtlggDrSTtKxflYApS:-KLtlggDrSTtKxflYApT:865845323) ([raw file](https://github.com/dgraph-io/gru/blob/82a2575cbbe34ea89251bd2bebfc096eca06f9bb/client/main.go#L176)):*
> ```Go
> 
> 				if count > 2 {
> 					// Show message (client can't reach server)
> ```

// TODO: Show message

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 1 of 2 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[client/main.go, line 129 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLtlt4j25H7Wkfp7LRi:-KLtpTgoem0IJc8_LNe_:182633054) ([raw file](https://github.com/dgraph-io/gru/blob/82a2575cbbe34ea89251bd2bebfc096eca06f9bb/client/main.go#L129)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> It's an unbuffered channel. This would definitely lead to blocking. If you're not sure why is it blocking, read the channels chapter in the book. You'll gain a good understanding of channels now that you're playing with them.
</details>
It would only be able to push to the channel when something is ready to consume from it. right? 

---

*[client/main.go, line 176 \[r5\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLtlggDrSTtKxflYApS:-KLtpvhb_-kf1T0ZvxnH:-499146009) ([raw file](https://github.com/dgraph-io/gru/blob/82a2575cbbe34ea89251bd2bebfc096eca06f9bb/client/main.go#L176)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> // TODO: Show message
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[client/main.go, line 118 \[r6\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLuJZh4hK18Ch3_G2uN:-KLuJZh4hK18Ch3_G2uO:-516057387) ([raw file](https://github.com/dgraph-io/gru/blob/6c53db14bc044ab7db0a50f3c8a19cad7b8198b4/client/main.go#L118)):*
> ```Go
> 	for {
> 		msg, err := stream.Recv()
> 		log.Println("got reply")
> ```

Since you are already logging error and `got reply. no error`, you could remove this.

---

*[client/main.go, line 129 \[r6\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLuJLkDtItlhHOu7Dji:-KLuJLkEuhfm8vWHqC8D:724495840) ([raw file](https://github.com/dgraph-io/gru/blob/6c53db14bc044ab7db0a50f3c8a19cad7b8198b4/client/main.go#L129)):*
> ```Go
> 
> func streamRecv(stream interact.GruQuiz_StreamChanClient) {
> 	gotPing := make(chan *interact.ServerStatus, 100)
> ```

Now, it might be better to name it msgChan :stuck_out_tongue:.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: 0 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[client/main.go, line 118 \[r6\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLuJZh4hK18Ch3_G2uN:-KLuaUC1d6LsAG4qDIcV:-499146009) ([raw file](https://github.com/dgraph-io/gru/blob/6c53db14bc044ab7db0a50f3c8a19cad7b8198b4/client/main.go#L118)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Since you are already logging error and `got reply. no error`, you could remove this.
</details>
Done.

---

*[client/main.go, line 129 \[r6\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLuJLkDtItlhHOu7Dji:-KLuaonkI7lgMP5jsU_0:-499146009) ([raw file](https://github.com/dgraph-io/gru/blob/6c53db14bc044ab7db0a50f3c8a19cad7b8198b4/client/main.go#L129)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Now, it might be better to name it msgChan :stuck_out_tongue:.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

pawanrawal commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[client/main.go, line 95 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLvEwSJ4_tJPfa7WbST:-KLvEwSJ4_tJPfa7WbSU:552941511) ([raw file](https://github.com/dgraph-io/gru/blob/f7776f3a24a72ea24767abc03e60c6d69e9f0957/client/main.go#L95)):*
> ```Go
> 			&interact.Req{Repeat: false, Sid: s.Id, Token: *token})
> 		try++
> 		if try > 100 {
> ```

Is this intentional? 

---

*[client/main.go, line 136 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLvF2ZUEFJROWRz35tH:-KLvF2ZUEFJROWRz35tI:988731526) ([raw file](https://github.com/dgraph-io/gru/blob/f7776f3a24a72ea24767abc03e60c6d69e9f0957/client/main.go#L136)):*
> ```Go
> }
> 
> func streamRecv(stream *interact.GruQuiz_StreamChanClient) {
> ```

Any particular reason to change it to a pointer?

---

*[client/main.go, line 162 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLvE2yLjqaWArKLb3Vu:-KLvE2yLjqaWArKLb3Vv:1310195721) ([raw file](https://github.com/dgraph-io/gru/blob/f7776f3a24a72ea24767abc03e60c6d69e9f0957/client/main.go#L162)):*
> ```Go
> 		case <-time.After(3 * PINGDUR):
> 			log.Println("No pings received from server")
> 			qp.pingbox.Text = "No pings from server"
> ```

We can change the message both in this case and the other case. Slack shows this. 
`Apologies. Trouble connecting to the server. Hang on for a sec?`

---

*[client/main.go, line 187 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLvEPMotryf_hr3tmmh:-KLvEPMotryf_hr3tmmi:1804933080) ([raw file](https://github.com/dgraph-io/gru/blob/f7776f3a24a72ea24767abc03e60c6d69e9f0957/client/main.go#L187)):*
> ```Go
> 
> 				if count > 2 {
> 					qp.pingbox.Text = "Can't ping server"
> ```

Also here. Do we show the final error page after waiting for x mins?

---

*[server/main.go, line 530 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLvEk6tRpnIpkrh-aM8:-KLvEk6tRpnIpkrh-aM9:-179631845) ([raw file](https://github.com/dgraph-io/gru/blob/f7776f3a24a72ea24767abc03e60c6d69e9f0957/server/main.go#L530)):*
> ```Go
> 					log.Printf("Stream: %v, sesion token: %v\n",
> 						err, c.sid)
> 					return
> ```

required?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->

ashwin95r commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[client/main.go, line 95 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLvEwSJ4_tJPfa7WbST:-KLvkXEYB-kxGcjhu5in:-1696250076) ([raw file](https://github.com/dgraph-io/gru/blob/f7776f3a24a72ea24767abc03e60c6d69e9f0957/client/main.go#L95)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Is this intentional? 
</details>
Yes. We don't want to show the error page too quickly

---

*[client/main.go, line 136 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLvF2ZUEFJROWRz35tH:-KLvkdUjCr_a0MBMOBFO:-2115166025) ([raw file](https://github.com/dgraph-io/gru/blob/f7776f3a24a72ea24767abc03e60c6d69e9f0957/client/main.go#L136)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Any particular reason to change it to a pointer?
</details>
So that we could assign a new stream object when the existing one gets stale when the server crashes.

---

*[client/main.go, line 162 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLvE2yLjqaWArKLb3Vu:-KLvkrmRbFKxC5330ZWc:-971703669) ([raw file](https://github.com/dgraph-io/gru/blob/f7776f3a24a72ea24767abc03e60c6d69e9f0957/client/main.go#L162)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> We can change the message both in this case and the other case. Slack shows this. 
> `Apologies. Trouble connecting to the server. Hang on for a sec?`
</details>
Sure, we could customise the message. But this would give more context on whether the client is unable to reach the server or the other way around.

---

*[client/main.go, line 187 \[r7\]](https://reviewable.io:443/reviews/dgraph-io/gru/19#-KLvEPMotryf_hr3tmmh:-KLvlC7Az-Z9wUQLGVry:1234551757) ([raw file](https://github.com/dgraph-io/gru/blob/f7776f3a24a72ea24767abc03e60c6d69e9f0957/client/main.go#L187)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary>
> Also here. Do we show the final error page after waiting for x mins?
</details>
We have to decide upon that. It is not shown now. But the user should be the judge of when he wants to close it as he loses a question when we show the error page and he has to restart the client.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/gru/19)*
<!-- Sent from Reviewable.io -->
